walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,121 @@
 
 package com.google.sps;
 
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Arrays;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+
+    private final ArrayList < TimeRange > options = new ArrayList < TimeRange > ();
+
+    public Collection < TimeRange > query(Collection < Event > events, MeetingRequest request) {
+
+        Collection attendees = request.getAttendees();
+        Collection optionalAttendees = request.getOptionalAttendees();
+
+        // check if duration is longer than the day. if so, return empty collection.
+        long duration = request.getDuration();
+        if (duration > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+
+        // if no attendees or optional attendees return collection of one item being the full day
+        if (attendees.isEmpty() && optionalAttendees.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+
+        // if there are optional attendees but no attendees, optional attendees will be treated as attendees
+        } else if (attendees.isEmpty()) {
+            attendees = optionalAttendees;
+        }
+
+        // if no events return the whole day
+        ArrayList < Event > eventsList = new ArrayList < Event > (events);
```

Comment:
why are you creating a copy of events here? by here, i mean "before it is needed". if event.isEmpty() is true, then your code will return immediately - and you have copied the events for no reason. prefer to avoid doing extra work like this unless it is necessary.

Reply:
done